### PR TITLE
[feat] add watsonx

### PR DESCRIPTION
**Pull request by h0rv**
### **User description**
This PR adds IBM watsonx.ai support via LiteLLM. Additionally, it adds support for Traceloop / OpenLLMetry via LiteLLM as well.

___

### **PR Type**
Enhancement, Configuration changes

___

### **Description**
- Added support for multiple Watsonx models in the initialization file.
- Integrated Traceloop and Watsonx in the LiteLLM AI handler.
- Updated temperature setting for chat completion across various tools.
- Added configuration templates for Traceloop and Watsonx in the secrets template.
- Added temperature configuration setting in the main configuration file.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Add support for Watsonx models in initialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/algo/__init__.py

- Added support for multiple Watsonx models.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-5923c546f24ec7308a0e43fc84bb6fe40de7bfe2ac6ee842da9578e5dc2c692b">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>litellm_ai_handler.py</strong><dd><code>Integrate Traceloop and Watsonx in LiteLLM AI handler</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/algo/ai_handlers/litellm_ai_handler.py

<li>Added Traceloop initialization.<br> <li> Added Watsonx environment variable setup.<br> <li> Configured max tokens for chat completion.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-ea1acaa0907f3410665530fbc4cda2ab524de2772e0bbe10bad4648b8be35dfe">+15/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_add_docs.py</strong><dd><code>Use configurable temperature for PR documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_add_docs.py

- Updated temperature setting for chat completion.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-0e80bd148ab0c27f12ee8f075c0089078acaa69021a44e4a09c916348033ab31">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_code_suggestions.py</strong><dd><code>Use configurable temperature for code suggestions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_code_suggestions.py

- Updated temperature setting for chat completion.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-b57ba775e741d6f80bc4f8154b71330c011dae0ac43f3d0197e785b3e6b7117b">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_description.py</strong><dd><code>Use configurable temperature for PR description</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_description.py

- Updated temperature setting for chat completion.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-66130451adce278a098a5770e2db52b12050205fcd4d80c13f813615d4449abb">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_generate_labels.py</strong><dd><code>Use configurable temperature for label generation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_generate_labels.py

- Updated temperature setting for chat completion.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-15739956ea92e33af13af3d1725ff1416bca4b087dadacb8aa1c1a9dd993f1c7">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_information_from_user.py</strong><dd><code>Use configurable temperature for user information</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_information_from_user.py

- Updated temperature setting for chat completion.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-4482d011ed824c14fee0398d9393bfeb3df899bbb328a2215fd66529920a733b">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_line_questions.py</strong><dd><code>Use configurable temperature for line questions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_line_questions.py

- Updated temperature setting for chat completion.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-7fc24c7ebf69233b72a5f56d003cc266ca0a952ecc7cf2e75123d0a1ab87d525">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_questions.py</strong><dd><code>Use configurable temperature for PR questions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_questions.py

- Updated temperature setting for chat completion.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-71970fafa37a18dfb27eb7b47754bfa7841e14686b181445298f914e03fc31aa">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_reviewer.py</strong><dd><code>Use configurable temperature for PR review</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_reviewer.py

- Updated temperature setting for chat completion.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-8e265068e189a06852605cc694b03e92b523b4f8162077a2f3455ced4cdad8dc">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_update_changelog.py</strong><dd><code>Use configurable temperature for changelog update</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/pr_update_changelog.py

- Updated temperature setting for chat completion.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-76d8133da16132fa69d308bf1e5edb4fc7de18343a4d99760960416c54c79087">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>.secrets_template.toml</strong><dd><code>Add Traceloop and Watsonx configuration templates</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/settings/.secrets_template.toml

<li>Added configuration for Traceloop.<br> <li> Added configuration for Watsonx.<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-f9cf1bb39fd1a7e126b4d06af0238b24125e183f540c20360a392ef695682783">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>configuration.toml</strong><dd><code>Add temperature configuration setting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/settings/configuration.toml

- Added temperature configuration.

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1047/files#diff-66cfda5143e484ee53ecf7aa0df7dca8ad0b181256f4b0675905db35bcbbae78">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by mrT23** [comment]
This PR contains multiple unrelated topics. I don't recommend it in general. Opening separate PRs per main topic will save time.

Some topics here will be approved, some require further discussion (and a new PR), and some will not:

(1) adding new Watson models - **approved**

(2) the change to `pr_agent/algo/utils.py` - **approved**

(3) adding **traceloop** - **not approved**.  there are man possible packages for logging. We won't include them in our docker, as they increase its size, which we really want to avoid (github action). This change is suitable for a **fork**.
(As a side note, even if we wanted this change, it would have required also significant documentation update)

(4) adding the temperature as a configuration option - in general, we want to avoid dead configuration options. If there is not a real convincing use case for a configuration, it should not exist.
If you can present a use case for why would changing the temperature is relevant to many users, I will consider it. If not, this change should be avoided.
In any case, it requires also a documentation update, and should be done in a separate PR

File: pr_agent/algo/ai_handlers/litellm_ai_handler.py [hunk 1]
```diff
@@ -126,6 +142,7 @@ async def chat_completion(self, model: str, system: str, user: str, temperature:
                 "deployment_id": deployment_id,
                 "messages": messages,
                 "temperature": temperature,
+                "max_tokens": get_settings().config.max_model_tokens,
```
Comment (mrT23) [inline pr_agent/algo/ai_handlers/litellm_ai_handler.py:145]: p.s.
this is wrong, and very dangerous.

if I understand correctly, litellm `max_tokens` is for the **output** `config.max_model_token` is something vet different

Comment (h0rv) [reply pr_agent/algo/ai_handlers/litellm_ai_handler.py]: fixed

**Comment by h0rv** [comment]
Hi @mrT23, thank you for the review. I reverted issues 3 and 4. Can you take another look please?

I will open an issue regarding the temperature configuration, as with newer range of models, Llama 3 405B, different hyperparameters can yield better performance from model to model.

File: pr_agent/algo/ai_handlers/litellm_ai_handler.py [hunk 2]
```diff
@@ -126,6 +135,7 @@ async def chat_completion(self, model: str, system: str, user: str, temperature:
                 "deployment_id": deployment_id,
                 "messages": messages,
                 "temperature": temperature,
+                "max_tokens": get_settings().config.max_tokens,
```
Comment (mrT23) [inline pr_agent/algo/ai_handlers/litellm_ai_handler.py:138]: what is this line ??

there is no such config parameter.

remove this

Comment (h0rv) [reply pr_agent/algo/ai_handlers/litellm_ai_handler.py:138]: you told me to change it [here](https://github.com/Codium-ai/pr-agent/pull/1047#discussion_r1687092276), as max_model_tokens wasn't right. max_model_tokens does not have have any docs. This is required for watsonx, as it only defaults to like 50 tokens for some reason.

I will make it watsonx specific

**Comment by mrT23** [comment]
i am closing this PR as it consistently contains a bug that will crush pr-agent ( asked 3 times to solve it !)

Reopen it only after this line is fully removed, and you have run, validated and triple checked it

**Comment by h0rv** [comment]
The `max_tokens` line is for a argument for LiteLLM, which supports this parameter: https://docs.litellm.ai/docs/providers/databricks#passing-additional-params---max_tokens-temperature

**Comment by h0rv** [comment]
@mrT23 I am sorry if I am missing something, but I do not understand what change would cause this to crash.

If necessary I can make the LiteLLM `max_tokens` param just for watsonx or add a new config to have it be set, but I believe this is a good change for the LiteLLM handler, as before the application configuration was not being used at all.

**Comment by mrT23** [comment]
watson models were added here in a clean PR:
https://github.com/Codium-ai/pr-agent/pull/1069

give the needed parameters as env variables

in addition, you can give custom models:
https://pr-agent-docs.codium.ai/usage-guide/changing_a_model/#custom-models

**Pull request closed without merging**